### ISSUE TITLE
FragmentStateChanger with show/hide does not trigger proper lifecycle

### ISSUE THREAD

**Issue by kakai248** [Zhuinden/simple-stack#133]
This is not an issue with the library, it's more of a heads up.

In the MVP samples, the state changer uses attach/detach. This called lifecycle methods properly (up to onStart/onStop) but will destroy the fragment view.

In the MVVM sample, the state changer uses show/hide. This won't destroy the fragment view but it also won't do any lifecycle at all.

This means that all the LiveData's of all the opened fragments will keep observing as usual.

I'm not sure how to avoid this because afaik there isn't a way to keep the view and do regular lifecycle. I'm not a fan of destroying the view because state will get lost. Navigation library does this and they refuse to allow show/hide.

**Comment by Zhuinden** [comment]
Hmm right LiveData would stay `active` even for fragments in the background.

But `detach` destroys the view hierarchy which can be expensive to create and noticeable while animating back.

Destroying the view with `detach` should not destroy the view because it does not call `onSaveInstanceState(Bundle` but it does call `view.saveHierarchyState` and preserves it.

Although if you're just observing *data* from a LiveData then receiving it even if you are not in foreground should not be a problem.

If you are observing commands like "please navigate" then that could be tricky.

I think removing and re-adding observers in `onHiddenChanged` would be excessive.

>  the state changer uses attach/detach.

ugh I should have changed this long ago

**Comment by kakai248** [comment]
> Although if you're just observing data from a LiveData then receiving it even if you are not in foreground should not be a problem.

It was with an event that I noticed this. I have a BehaviorRelay that notifies when a model gets deleted. I have two screens (in the stack) that upon receiving this event should close. My events are propagated to the Fragment through a `LiveData<Event<Something>>`. This `Event` class makes sure events are only handled once. The first fragment receiving the event was actually the oldest one, which means that it will actually pop the newer fragment. So all in all, only one fragment gets popped.

I would have to build another system to propagate events instead of relying on LiveData. It would need a different lifecycle, which makes code hard to follow. But the main issue is still there and it's easy to fall into some sort of error because of this.

Fragments and their goddamn lifecycles -_-

**Comment by Zhuinden** [comment]
I think LiveData is a decent replacement for BehaviorRelays, but trying to replace either PublishRelay or DispatchWorkSubject with LiveData is just asking for trouble.

Either way it kinda sucks that you are correct and you'd need a custom fragment lifecycle on top of the already existing fragment *and view* lifecycles that is also made inactive by `onHiddenChanged(false)`.

**Comment by Zhuinden** [comment]
@yigit that makes me wonder though, any particular reason why `fragment.isHidden() == true` does not mean its Lifecycle is `inactive` by default?

**Comment by matejdro** [comment]
How about something like that? https://gist.github.com/matejdro/4b2323158a5fd2b335c56500a9fee374

Basically we could create our own `LifecycleOwner` to use for LiveData that takes visibility into account. When calling `LiveData.observe()`, you would just pass `visibleViewLifecycle` as lifecycle owner.

**Comment by Zhuinden** [comment]
The real magic with Fragments is that `onStop` is triggered when it is detached, even though the Activity around it is not stopped... Indeed, a combination of onHiddenChanged(true) or onStopped means that you're not in front.

I can't say anything about `setUserVisibilityHint`.

**Comment by matejdro** [comment]
> `onStop` is triggered when it is detached, even though the Activity around it is not stopped

I would argue that this is correct behavior, since Fragment itself IS stopped.

**Comment by Zhuinden** [comment]
So what of hidden? :D

**Comment by matejdro** [comment]
Well, the issue is that even though fragment is no longer visible to the user, it does not trigger `onStop` by just hiding.

**Comment by matejdro** [comment]
I think my current solution for this issue will be following:

1. Keep using hide/show version because going back is much smoother
2. Add `fragment.userVisibleHint = true/false` after every `fragmentTransaction.show/hide` call inside `FragmentStateChanger`. I'm not sure if that even helps with anything, but I guess it can't hurt.
3. Use https://gist.github.com/matejdro/4b2323158a5fd2b335c56500a9fee374 as base class for all fragments. This base provides custom lifecycle for LiveData and additional `onFragmentDisplayed` and `onFragmentHidden` callbacks that mimic existing `onStart/Stop` callbacks, but take visibility into acount.

**Comment by Zhuinden** [comment]
@matejdro you could make onStart/onStop final in your base fragment.

**Comment by Zhuinden** [comment]
Either way, I'm not really using `SingleLiveEvent` in these samples, and this is about fragment intricacies.

I think I'll close it.

**Comment by matejdro** [comment]
Maybe a warning or notice somewhere in samples would be in order? People generally expect for Fragment lifecycles to happen, that is the whole point of Fragments.

**Comment by Zhuinden** [comment]
Sigh, I guess I'll have to move the samples back to `detach/attach` again, although the layout inflation cost on back navigation can be noticeable.

IRL we used `show/hide` and it worked for us just fine. Honestly, the FragmentStateChanger you see here is very close to what we were doing apart from additionally keeping some other fragments added but hidden (for the bottom nav to be fast).

**Comment by matejdro** [comment]
Yes, `show/hide` works fine, as long as you expect and know that lifecycle methods will not happen. I think leaving it at `show/hide` is fine, as long as it is noted in documentation.

**Comment by matejdro** [comment]
It looks like this situation may be improved in future releases: https://issuetracker.google.com/issues/129780800

**Comment by Zhuinden** [comment]
Lipstick on a pig... but it will definitely make it easier to combine the hidden state into onStop.

Good idea when it comes out :+1:

**Comment by kakai248** [comment]
Can confirm `.setMaxLifecycle` fixes this.

Change all shows to:
```kotlin
fragmentTransaction
    .show(fragment)
    .setMaxLifecycle(fragment, Lifecycle.State.RESUMED)
```

And all hides to:
```kotlin
fragmentTransaction
    .hide(fragment)
    .setMaxLifecycle(fragment, Lifecycle.State.CREATED)
```

**Comment by Zhuinden** [comment]
This requires `Version 1.1.0-alpha07`, correct?

I'll update the samples when I get to it, but I'm happy to hear it works. This `setMaxLifecycle` is tricky, but at least it works well!

**Comment by matejdro** [comment]
Is `Lifecycle.State.RESUMED` enough to affect LiveData? If I recall correctly, fragment must not be started, not just not-resumed.

**Comment by kakai248** [comment]
> This requires androidx.fragment:fragment:1.1.0-alpha07, correct?

Yes

> Is Lifecycle.State.RESUMED enough to affect LiveData? If I recall correctly, fragment must not be started, not just not-resumed.

RESUMED removes the restriction so the fragment can reach onResume again. 

CREATED doesn't let the fragment be on a started state so it's exactly what we need for LiveData. I tested it and you can see onStop being called.

**Comment by kakai248** [comment]
In fragments version 1.1.0-alpha08, they changed the behavior of `setMaxLifecycle(fragment, Lifecycle.State.CREATED)` to destroy the view. [See this commit](https://android-review.googlesource.com/c/platform/frameworks/support/+/954180/)

This means that using show/hide in combination with `setMaxLifecycle` will no longer keep the view, which makes it worthless. We're back to where we started.

Either we use show/hide and lose lifecycle or just replace it altogether with attach/detach. We went with the latter.

**Comment by Zhuinden** [comment]
Disappointing. No idea what to do about it then.

At this point, the solution might be to just not use fragments and use Views or Conductor instead. 🤔

**Comment by matejdro** [comment]
@kakai248 Did you link to the right commit? The one you linked just seems to prevent fragments from stopping at ACTIVITY_CREATED.

**Comment by kakai248** [comment]
@matejdro It means that they will no longer stop at `ACTIVITY_CREATED` and will go all the way back to `ON_CREATED`. According to [fragment lifecycle](https://static.javatpoint.com/images/androidimages/fragmentlifecycle.png), it means the view will have to be destroyed so it can reach that state.

Unless I'm mistaken. We were actually debugging other issue with the fragment view and started seeing a flicker after updating the dependency. Upon checking, `onDestroyView` was getting called, which was not supposed with show/hide.

**Comment by Zhuinden** [comment]
They should just not re-use `Lifecycle.State` for this, because it doesn't match the Fragment lifecycle.

**Comment by matejdro** [comment]
Maybe we should make an issue on issue tracker for this? Who knows, maybe planets align and they actually see the issue.

**Comment by Zhuinden** [comment]
> Bug fixes
>
> - When using setMaxLifecycle(Lifecycle.State.CREATED), Fragments will properly have their view destroyed. [aosp/954180](https://android-review.googlesource.com/c/platform/frameworks/support/+/954180/)

:roll_eyes:

**Comment by matejdro** [comment]
I have created ticket for that: https://issuetracker.google.com/issues/133100471

**Comment by matejdro** [comment]
So, they closed the issue:

> This is working as intended in the current world of Fragments.

> The creation of a Fragment's view happens immediately before being started and cases where the Fragment is stopped, such as going onto the back stack, result in the Fragment's view being destroyed.

> We'll be working to combine the view and fragment lifecycles as part of https://issuetracker.google.com/issues/127528777

Maybe this combining of lifecycles could work in our favor?

**Comment by Zhuinden** [comment]
Can barely wait for them to depreciate `setRetainInstance` for that, because that's the only way to do that :thinking_face:

Let's hope they don't start throwing exceptions for using platform level fragments. This wouldn't be the first time they break backwards compatibility in such a way (see region ops).

Either way, that means you really do need to define a custom visibility LifecycleOwner. :disappointed:

**Comment by matejdro** [comment]
Hey, at least the future is bright:

> Once you opt in to combining the lifecycle (that future set of work linked in #4), Lifecycle.State.CREATED will mean that both the Fragment is created and the Fragment's View is created. That'll apply to all parts of FragmentManager including setMaxLifecycle().

**Comment by Zhuinden** [comment]
@kakai248 https://twitter.com/ianhlake/status/1188821956033835008 🤔

**Comment by kakai248** [comment]
Tbh with view recreation everything runs flawlessly (at least on my device, which is a fast one). And I take a simplification of the fragment lifecycle over keeping views any day.

**Comment by Zhuinden** [comment]
It's kind of a pain if you have something heavier like a MapFragment for example, though.

**Comment by kakai248** [comment]
That probably can be solved by AsyncLayoutInflater and MapView? Never tried it though. I'm not using maps extensively.

**Comment by matejdro** [comment]
Would you even want to have MapView loaded in the background? It sounds like a lot of memory wasted just sitting in the backstack.

**Comment by Zhuinden** [comment]
@matejdro sounds like a better UX than waiting for it to reload each time though, IIRC it loads for like 4 seconds.

**Comment by Zhuinden** [comment]
In the end, this all boils down to that Google does not want to support visibility-based lifecycle owners, they want to kill `attach/detach` which might be quite awkward if you think about it (how will they know whether the Fragment state should survive? Are we to juggle using `setInitialSavedState`? how will it keep the ViewModelStore's ViewModel?).

I guess that's one more reason to use Rx instead of LiveData :wink:

**Comment by methelas** [comment]
I know that this issue is closed for a while now... but what do you think about this solution to solve the hide/show lifecycle issue? It provides both regular and view-specific lifecycle owners that are aware of the `isHidden` status. It also checks whether any parent fragment is hidden too(if youre using nested fragments) since `onHiddenChaged` is not called for nested fragments while hiding parent fragment.
I think I'm a little bit over my head here, so if you see any issue with it, please let me know 🙈 

```kotlin
abstract class BaseFragment : Fragment() {

    //  LIFECYCLE OWNERS
    private val _hideableLifecycleOwner = HideableLifecycleOwner()
    val hideableLifecycle: LifecycleOwner get() = _hideableLifecycleOwner
    private val _hideableViewLifecycleOwner = HideableViewLifecycleOwner()
    val hideableViewLifecycle: LifecycleOwner get() = _hideableViewLifecycleOwner

    //  HIDDEN
    private var onHiddenOrParentHiddenChangedCalled = true
    private val isHiddenOrParentHidden: Boolean
        get() = isHidden || ((parentFragment as BaseFragment?)?.isHiddenOrParentHidden == true)
    private val _isHiddenForChildren = MutableLiveData(false)
    val isHiddenForChildren: LiveData<Boolean> = _isHiddenForChildren.distinctUntilChanged()

    // ===== ===== ===== ===== ===== ===== ===== ===== ===== ===== ===== ===== ===== ===== ===== ===
    // LIFECYCLE
    // ===== ===== ===== ===== ===== ===== ===== ===== ===== ===== ===== ===== ===== ===== ===== ===

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        //  HIDDEN
        /* start listening to hide state changes of your direct parent(if exists) */
        (parentFragment as BaseFragment?)?.isHiddenForChildren?.observe(this, Observer {
            invokeOnHiddenOrParentHiddenChanged()
        })
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        //  LIFECYCLE OWNERS
        _hideableViewLifecycleOwner.initialize()
    }

    override fun onDestroyView() {
        //  LIFECYCLE OWNERS
        _hideableViewLifecycleOwner.deinitialize()
        super.onDestroyView()
    }

    override fun onDestroy() {
        //  HIDDEN
        (parentFragment as BaseFragment?)?.isHiddenForChildren?.removeObservers(this)
        super.onDestroy()
    }

    // ===== ===== ===== ===== ===== ===== ===== ===== ===== ===== ===== ===== ===== ===== ===== ===
    // HIDDEN
    // ===== ===== ===== ===== ===== ===== ===== ===== ===== ===== ===== ===== ===== ===== ===== ===

    final override fun onHiddenChanged(hidden: Boolean) {
        super.onHiddenChanged(hidden)
        invokeOnHiddenOrParentHiddenChanged()
    }


    private fun invokeOnHiddenOrParentHiddenChanged() {
        onHiddenOrParentHiddenChangedCalled = false
        onHiddenOrParentHiddenChanged(isHiddenOrParentHidden)
        if (!onHiddenOrParentHiddenChangedCalled)
            throw IllegalStateException("super.onHiddenOrParentHiddenChanged() not called!")
    }

    protected open fun onHiddenOrParentHiddenChanged(hidden: Boolean) {
        _hideableLifecycleOwner.onHiddenChanged(hidden)
        _hideableViewLifecycleOwner.onHiddenChanged(hidden)
        _isHiddenForChildren.value = hidden
        onHiddenOrParentHiddenChangedCalled = true
    }

    // ===== ===== ===== ===== ===== ===== ===== ===== ===== ===== ===== ===== ===== ===== ===== ===
    // ===== ===== ===== ===== ===== ===== ===== ===== ===== ===== ===== ===== ===== ===== ===== ===
    // LIFECYCLE OWNERS
    // ===== ===== ===== ===== ===== ===== ===== ===== ===== ===== ===== ===== ===== ===== ===== ===
    // ===== ===== ===== ===== ===== ===== ===== ===== ===== ===== ===== ===== ===== ===== ===== ===

    private inner class HideableLifecycleOwner : LifecycleOwner, LifecycleEventObserver {
        private val lifecycleRegistry = LifecycleRegistry(this)

        init {
            this@BaseFragment.lifecycle.addObserver(this)
        }

        override fun onStateChanged(source: LifecycleOwner, event: Lifecycle.Event) {
            when (event) {
                Lifecycle.Event.ON_START, Lifecycle.Event.ON_RESUME,
                Lifecycle.Event.ON_PAUSE, Lifecycle.Event.ON_STOP ->
                    lifecycleRegistry.handleLifecycleEvent(if (this@BaseFragment.isHiddenOrParentHidden) Lifecycle.Event.ON_STOP else event)
                else -> lifecycleRegistry.handleLifecycleEvent(event)
            }
        }

        internal fun onHiddenChanged(hidden: Boolean) {
            when {
                this@BaseFragment.lifecycle.currentState.isAtLeast(Lifecycle.State.STARTED) && hidden ->
                    lifecycleRegistry.handleLifecycleEvent(Lifecycle.Event.ON_STOP)
                this@BaseFragment.lifecycle.currentState.isAtLeast(Lifecycle.State.RESUMED) -> lifecycleRegistry.handleLifecycleEvent(Lifecycle.Event.ON_RESUME)
                this@BaseFragment.lifecycle.currentState.isAtLeast(Lifecycle.State.STARTED) -> lifecycleRegistry.handleLifecycleEvent(Lifecycle.Event.ON_START)
                this@BaseFragment.lifecycle.currentState.isAtLeast(Lifecycle.State.CREATED) -> lifecycleRegistry.handleLifecycleEvent(Lifecycle.Event.ON_CREATE)
            }
        }

        override fun getLifecycle(): Lifecycle = lifecycleRegistry
    }

    private inner class HideableViewLifecycleOwner : LifecycleOwner, LifecycleEventObserver {
        private var isInitialized = false
        private val lifecycleRegistry = LifecycleRegistry(this)

        fun initialize() {
            this@BaseFragment.viewLifecycleOwner.lifecycle.addObserver(this)
            isInitialized = true
        }

        fun deinitialize() {
            isInitialized = false
            this@BaseFragment.viewLifecycleOwner.lifecycle.removeObserver(this)
            lifecycleRegistry.handleLifecycleEvent(Lifecycle.Event.ON_DESTROY)
        }

        override fun onStateChanged(source: LifecycleOwner, event: Lifecycle.Event) {
            when (event) {
                Lifecycle.Event.ON_START, Lifecycle.Event.ON_RESUME,
                Lifecycle.Event.ON_PAUSE, Lifecycle.Event.ON_STOP ->
                    lifecycleRegistry.handleLifecycleEvent(if (this@BaseFragment.isHiddenOrParentHidden) Lifecycle.Event.ON_STOP else event)
                else -> lifecycleRegistry.handleLifecycleEvent(event)
            }
        }

        internal fun onHiddenChanged(hidden: Boolean) {
            if (!isInitialized) return
            when {
                this@BaseFragment.viewLifecycleOwner.lifecycle.currentState.isAtLeast(Lifecycle.State.STARTED) && hidden ->
                    lifecycleRegistry.handleLifecycleEvent(Lifecycle.Event.ON_STOP)
                this@BaseFragment.viewLifecycleOwner.lifecycle.currentState.isAtLeast(Lifecycle.State.RESUMED) -> lifecycleRegistry.handleLifecycleEvent(Lifecycle.Event.ON_RESUME)
                this@BaseFragment.viewLifecycleOwner.lifecycle.currentState.isAtLeast(Lifecycle.State.STARTED) -> lifecycleRegistry.handleLifecycleEvent(Lifecycle.Event.ON_START)
                this@BaseFragment.viewLifecycleOwner.lifecycle.currentState.isAtLeast(Lifecycle.State.CREATED) -> lifecycleRegistry.handleLifecycleEvent(Lifecycle.Event.ON_CREATE)
            }
        }

        override fun getLifecycle(): Lifecycle = lifecycleRegistry
    }
}
```

**Comment by matejdro** [comment]
This could work, but even if you do it, `hide()` and `show() will be removed some time in the future, so you would be building your architure on already deprecated methods.

**Comment by Zhuinden** [comment]
Technically they want to kill `attach` and `detach` too, because Ian Lake doesn't like them.

And whatever Navigation wants, Navigation gets.

Although IIRC it is possible to simulate `detach/attach` with `setMaxLifecycle(CREATED)`, so not all hope is lost.

As for that solution above, I'll try to look through it when I'm near a PC. I'm sure it can be simplified a bit using `viewLifecycleOwnerLiveData`.

**Comment by methelas** [comment]
@Zhuinden I mean... if they really getting rid of hide/show then I just wasted my time trying to figure it out 🤔 
Should I just go for `attach/detach` approach for now?
And when they deprecate them too, will SimpleStack adjust?

**Comment by Zhuinden** [comment]
>  if they really getting rid of hide/show then I just wasted my time trying to figure it out 

> Should I just go for attach/detach approach for now?

If you want to keep the view hierarchy as you navigate (which you probably do because re-inflation is costly), then you should keep your new `show/hide` approach :slightly_smiling_face: 

> And when they deprecate them too, will SimpleStack adjust?

Simple-Stack *itself* doesn't care as it is completely independent of whether you use Fragments, Views, or Conductor Controllers, or something completely different. The backstack / navigator do not depend on any behavior from the support library.

The *samples* will most likely need to adjust, although for the samples to adjust, Simple-Stack needs to shift to use AndroidX (currently I use the nullability annotations and those are still from `android.support.*`). I'd have to replace `detach/attach` with `setMaxLifecycle(CREATED/RESUMED`.

**Comment by methelas** [comment]
@Zhuinden 
Thanks. I guess I'll go with hide/show for now, since at this point the app is pretty simple anyway.
And when the time comes I'll change it to more up-to-date solution.
